lmdeploy/pytorch_poc/patch/baichuan.py

```diff
@@ -0,0 +1,327 @@
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor],
+               Optional[Tuple[torch.Tensor]]]:
+        assert not output_attentions
+        origin_self = self.origin_mod
+
+        context = self.context.context
+        history_lengths = context.history_lengths
+        max_seq_len = position_ids.size(-1)
+
+        proj = origin_self.W_pack(hidden_states)
+        proj = proj.unflatten(
+            -1, (3, origin_self.hidden_size)).unsqueeze(0).transpose(
+                0, -2).squeeze(-2)
+        query_states = proj[0].view(-1, origin_self.num_heads,
+                                    origin_self.head_dim)
+        key_states = proj[1].view(-1, origin_self.num_heads,
+                                  origin_self.head_dim)
+        value_states = proj[2].view(-1, origin_self.num_heads,
+                                    origin_self.head_dim)
+
+        kv_seq_len = max_seq_len + max(history_lengths)
+        # TODO: setup past_key_value with paged attention
+        if hasattr(origin_self,
+                   'rotary_emb'):  # baichuan-13B has no rotary_emb
+            cos, sin = origin_self.rotary_emb(value_states, seq_len=kv_seq_len)
+            query_states, key_states = apply_rotary_pos_emb(
+                query_states, key_states, cos, sin, position_ids)
+
+        kv_seq_length = position_ids[..., -1] + 1
+        q_seq_length = kv_seq_length - kv_seq_length.new_tensor(
+            history_lengths)
+        q_start_loc = q_seq_length.cumsum(0)
+        q_start_loc = torch.cat([q_start_loc.new_zeros(1), q_start_loc[:-1]])
+        context.fill_cache(
+            key_states,
+            value_states,
+            q_start_loc,
+            q_seq_length,
+            past_key_value[0],
+            past_key_value[1],
+        )
+        attn_output = torch.empty_like(query_states)
+
+        block_offsets = context.block_offsets
+        block_size = past_key_value[0].size(1)
+        paged_attention_fwd(query_states,
+                            past_key_value[0],
+                            past_key_value[1],
+                            attn_output,
+                            block_offsets,
+                            b_start_loc=q_start_loc,
+                            b_seq_len=q_seq_length,
+                            b_kv_seq_len=kv_seq_length,
+                            max_input_len=max_seq_len,
+                            BLOCK=block_size)
+        attn_output = attn_output.reshape(-1, origin_self.hidden_size)
+
+        attn_output = origin_self.o_proj(attn_output)
+
+        if not output_attentions:
+            attn_weights = None
+
+        return attn_output, attn_weights, past_key_value
+
+
+class BaichuanLayer(torch.nn.Module):
```

Comment:
It seems this modification is only for allowing position ids passed into attention to compute q_locs and q_seq_lens for paged attention. Since these two are crucial for paged attention, so I think we can pass these two values consistently through the `context`.to avoid copying/pasting a lot of codes just for changing the argument list.
I have made this work in the chatglm part, https://github.com/InternLM/lmdeploy/pull/360/files#diff-a5adea9960ed4a4cc81eee9f97a1bdb9e8353832cdfc3d56a301a840d35a545bR490. If all (cc @grimoire ) think this is a good idea, I can make a standalone PR for this.

Reply:
https://github.com/InternLM/lmdeploy/pull/398